source/common/quic/platform/quiche_flags_impl.h

```diff
@@ -76,55 +77,47 @@ class Flag {
 template <typename T> class TypedFlag : public Flag {
 public:
   TypedFlag(const char* name, T default_value, const char* help)
-      : Flag(name, help), value_(default_value), default_value_(default_value) {}
+      : Flag(name, help), value_(default_value), explicit_value_(default_value), default_value_(default_value) {
+    ASSERT(std::atomic_is_lock_free(&value_));
+  }
 
   bool setValueFromString(const std::string& value_str) override;
 
   void resetValue() override {
-    absl::MutexLock lock(&mutex_);
-    value_ = default_value_;
+    explicit_value_.store(default_value_, std::memory_order_relaxed);
+    value_.store(default_value_, std::memory_order_relaxed);
   }
 
   // Set flag value.
   void setValue(T value) {
-    absl::MutexLock lock(&mutex_);
-    value_ = value;
+    explicit_value_.store(value, std::memory_order_relaxed);
+    value_.store(value, std::memory_order_relaxed);
   }
 
   // Return flag value.
   T value() const {
-    absl::MutexLock lock(&mutex_);
-    if (has_reloaded_value_) {
-      return reloaded_value_;
-    }
     return value_;
   }
 
   void setReloadedValue(T value) {
-    absl::MutexLock lock(&mutex_);
-    has_reloaded_value_ = true;
-    reloaded_value_ = value;
+    value_.store(value, std::memory_order_relaxed);
   }
 
   void resetReloadedValue() override {
-    absl::MutexLock lock(&mutex_);
-    has_reloaded_value_ = false;
+    value_.store(explicit_value_, std::memory_order_relaxed);
   }
 
 private:
-  mutable absl::Mutex mutex_;
-  T value_ ABSL_GUARDED_BY(mutex_);
+  std::atomic<T> value_;
+  std::atomic<T> explicit_value_;
```

Comment:
Can we comment these up?  Alternately if using absl flags works, we might just want to migrate straight over :-)
/wait

Reply:
Added comments. I'm working on the absl flag approach now but it's going to take a bit of work to wrap the absl::Flags in the QUICHE flag api, and refactor things accordingly. Once that's working I'll work with the internal folks to get a max QPS load test and validate the performance properties. There is a comment about `GetFlag()` which says not to use it in tight loops, so I'm very curious to understand the perf impact.

https://github.com/abseil/abseil-cpp/blob/master/absl/flags/flag.h#L84

Reply:
I've converted this to a draft and will revisit after we address the various other flag cleanup/refactor PRs get sorted out.